bowseries.go

```diff
@@ -19,85 +19,100 @@ type Series struct {
+					}, nil, length-bitutil.CountSetBits(valid, 0, length), 0),
+			),
+		}
+	case Boolean:
+		mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
+		builder := array.NewBooleanBuilder(mem)
+		defer builder.Release()
+		builder.AppendValues(dataArray.([]bool), buildNullBitmapBool(len(dataArray.([]bool)), validityArray))
+		return Series{Name: name, Array: builder.NewArray()}
+	case String:
+		mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
+		builder := array.NewStringBuilder(mem)
+		defer builder.Release()
+		builder.AppendValues(dataArray.([]string), buildNullBitmapBool(len(dataArray.([]string)), validityArray))
+		return Series{Name: name, Array: builder.NewArray()}
 	default:
-		panic(fmt.Errorf("unsupported type %T", v))
+		panic(fmt.Errorf("unsupported type '%s'", typ))
 	}
 }
 
 // NewSeriesFromBuffer returns a new Series from a name and a Buffer.
 func NewSeriesFromBuffer(name string, buf Buffer) Series {
-	switch data := buf.Data.(type) {
-	case []int64:
-		return newInt64Series(name, data, buf.nullBitmapBytes)
-	case []float64:
-		return newFloat64Series(name, data, buf.nullBitmapBytes)
-	case []bool:
-		return newBooleanSeries(name, data, buf.nullBitmapBytes)
-	case []string:
-		return newStringSeries(name, data, buf.nullBitmapBytes)
+	switch buf.DataType {
+	case Int64:
+		length := len(buf.Data.([]int64))
+		return Series{
+			Name: name,
+			Array: array.NewInt64Data(
+				array.NewData(mapBowToArrowTypes[Int64], length,
+					[]*memory.Buffer{
+						memory.NewBufferBytes(buf.nullBitmapBytes),
+						memory.NewBufferBytes(arrow.Int64Traits.CastToBytes(buf.Data.([]int64))),
+					}, nil, length-bitutil.CountSetBits(buf.nullBitmapBytes, 0, length), 0),
+			),
+		}
+	case Float64:
+		length := len(buf.Data.([]float64))
+		return Series{
+			Name: name,
+			Array: array.NewFloat64Data(
+				array.NewData(mapBowToArrowTypes[Float64], length,
+					[]*memory.Buffer{
+						memory.NewBufferBytes(buf.nullBitmapBytes),
+						memory.NewBufferBytes(arrow.Float64Traits.CastToBytes(buf.Data.([]float64))),
+					}, nil, length-bitutil.CountSetBits(buf.nullBitmapBytes, 0, length), 0),
+			),
+		}
+	case Boolean:
+		mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
+		builder := array.NewBooleanBuilder(mem)
+		defer builder.Release()
+		builder.AppendValues(buf.Data.([]bool), buildNullBitmapBool(len(buf.Data.([]bool)), buf.nullBitmapBytes))
+		return Series{Name: name, Array: builder.NewArray()}
+	case String:
+		mem := memory.NewCheckedAllocator(memory.NewGoAllocator())
+		builder := array.NewStringBuilder(mem)
+		defer builder.Release()
+		builder.AppendValues(buf.Data.([]string), buildNullBitmapBool(len(buf.Data.([]string)), buf.nullBitmapBytes))
+		return Series{Name: name, Array: builder.NewArray()}
 	default:
-		panic(fmt.Errorf("unsupported type '%T'", buf.Data))
+		panic(fmt.Errorf("unsupported type '%s'", buf.DataType))
 	}
 }
 
-func newInt64Series(name string, data []int64, valid []byte) Series {
```

Comment:
why remove the functions since they are called twice already ? Seems like an unnecessary change ;)
Doesn't matter much, you can resolve this.

Reply:
I just improved this ;)